src/store_client.cc

```diff
@@ -146,22 +146,42 @@ storeClientListAdd(StoreEntry * e, void *data)
     return sc;
 }
 
+/// schedules asynchronous STCB call to relay disk or memory read results
+/// \param outcome an error signal (if negative), an EOF signal (if zero), or the number of bytes read
 void
-store_client::callback(ssize_t sz, bool error)
+store_client::callback(const ssize_t outcome)
 {
-    size_t bSz = 0;
+    if (outcome > 0)
+        return noteCopiedBytes(outcome);
 
-    if (sz >= 0 && !error)
-        bSz = sz;
+    if (outcome < 0)
+        return fail();
 
-    StoreIOBuffer result(bSz, 0,copyInto.data);
+    noteEof();
+}
 
-    if (sz < 0 || error)
-        result.flags.error = 1;
+/// finishCallback() wrapper; TODO: Add NullaryMemFunT for non-jobs.
+void
+store_client::FinishCallback(store_client * const sc)
+{
+    sc->finishCallback();
+}
 
-    result.offset = cmp_offset;
-    assert(_callback.pending());
-    cmp_offset = copyInto.offset + bSz;
+/// finishes a copy()-STCB sequence by synchronously calling STCB
+void
+store_client::finishCallback()
+{
+    Assure(_callback.callback_handler);
+    Assure(_callback.notifier);
+
+    // callers are not ready to handle a content+error combination
+    Assure(object_ok || !copiedSize);
+
+    StoreIOBuffer result(copiedSize, copyInto.offset, copyInto.data);
+    result.flags.error = object_ok ? 0 : 1;
+    copiedSize = 0;
+
+    cmp_offset = copyInto.offset + copyInto.length;
```

Comment:
I guess we should calculate this variable as before - adding up the number of copied bytes to offset instead of buffer length:
```suggestion
    cmp_offset = copyInto.offset + copiedSize;
    copiedSize = 0;
```

Reply:
Good catch! I committed a fix: https://github.com/measurement-factory/squid/commit/297053e31205dd56c7ffcccc3fda916b0e8fc46c

I meant to say result.length (which is also copiedSize).

We could use copiedSize to reduce word diff, but perhaps the committed version is a bit clearer.